app/src/main/kotlin/com/google/samples/apps/nowinandroid/NiaApplication.kt

```diff
@@ -41,7 +40,7 @@ class NiaApplication : Application(), ImageLoaderFactory {
         // Initialize Sync; the system responsible for keeping data in the app up to date.
         Sync.initialize(context = this)
         profileVerifierLogger()
+        // We set immediately Coil's image loader factory to prevent initialization with the first image.
+        Coil.setImageLoader(imageLoaderAsyncFactory)
```

Comment:
I don't actually understand why you need this call, and to remove ImageLoaderFactory.

The `imageLoaderAsyncFactory` will cause your early init of the loader. The call to newImageLoader() could call your AsyncFactory.

Reply:
If we leave it as it was, Coil will request the `newImageLoader` with the first loading image on the main thread, which I think is way too late. There's a huge time window we can initialize it before that happens (and on a background thread).

Reply:
It doesn't seem to change timing. Injection will cause it to kick that off at same point in time.

Reply:
I guess I misunderstood what you meant. Yes, we could keep this class implementing the `ImageLoagerFactory` and inject the `imageLoaderAsyncFactory`. I thought to change it because someone might ask why we have 2 classes implementing `ImageLoaderFactory`. 

This way I thought it's more clear that we set the factory immediatelly, but it is doing async init internally.

I also thought about the fact that when dagger injects the `imageLoaderAsyncFactory` it comes with a side effect of starting the initialization (in the `init` of the class). I was thinking this is not great and had a `initialize()` function as well, which was called in `Application.onCreate` .. but then I realised I just overcomplicate it.

I can change it back :D

Reply:
Your call.

Reply:
Why not upstream this to Coil cleanly? Update the sample to use a provider, avoid early init.

Have a simple method to kick off init in background? Perhaps conditionally, say based on process lifecycle in case a service is run?